### ISSUE TITLE
TestContext lack of thread-safety breaks tests classes using fluent API running in parallel

### ISSUE THREAD

**Issue by Eldar1205** [TestStack/TestStack.BDDfy#222]
I tried running xUnit tests with BDDfy's fluent API. When I ran each test class on its own everything worked, but when running several test classes in parallel there were non-deterministic errors coming from TestContext class. I examined the code and saw there is a static mutable dictionary inside the class, accessed without locks. 

Possible solutions:
1. Use ConcurrentDictionary, its semantics are exactly what TestContext needs (preferred, but requires .Net 4 and above)
2. Use locks when accessing the dictionary (.Net 2 and above)

**Comment by mwhelan** [comment]
I think ConcurrentDictionary might be a good solution. What do people think of raising BDDfy to .Net 4?

**Comment by robdmoore** [comment]
Most things seem to support 4.5 only these days, bumnping to .NET 4 is fine, you might just want to bump major version.

**Comment by mwhelan** [comment]
So, BDDfy 5 then?

Do you think we should go to 4.5 then, or stick to 4?

How about Microsoft ending support for .NET Framework 4, 4.5 and 4.5.1? Should we maybe go to 4.5.2?

**Comment by Eldar1205** [comment]
Hi,

I'm the original issuer, also from Microsoft, and I can tell you that
Microsoft isn't ending support for those frameworks. We in fact still use
.Net 4.5.1 internally. Most moved to .Net 4.5 to enjoy the async/await
goodies.

Thanks for your fast reponse

2016-01-31 16:09 GMT+02:00 Michael Whelan <notifications@github.com>:

**Comment by mwhelan** [comment]
Oh? I was going by this blog post:
http://blogs.msdn.com/b/dotnet/archive/2015/12/09/support-ending-for-the-net-framework-4-4-5-and-4-5-1.aspx

**Comment by Eldar1205** [comment]
I'm aware of that, and still there are some developers that won't do that
upgrade for coming time, including myself and I'd really like to keep using
BDDfy and receive the fix.

2016-01-31 16:26 GMT+02:00 Michael Whelan <notifications@github.com>:

**Comment by mwhelan** [comment]
Yes, we want to support as many people as we can so no need to upgrade any higher than the features we need. I'm happy to move to .Net 4 and the ConcurrentDictionary that you suggested.

Is this change to TestContext with ConcurrentDictionary something you might be able to submit a PR for?

**Comment by Eldar1205** [comment]
I'm waiting for response from Lawyers about contributing to OSS on behalf
of Microsoft, unfortunate legal issues. If they won't response in the next
few days I'll submit a PR myself, though it'll be my first time
contributing to an OSS using Github so it might take a while to get right.

2016-01-31 16:44 GMT+02:00 Michael Whelan <notifications@github.com>:

**Comment by mwhelan** [comment]
Fair enough. Well, I'm keen to try to solve the problem as parallel testing is an important use case. 

Could you possibly provide a test case or two that illustrates the non-deterministic errors you were getting from the TestContext class?

**Comment by Eldar1205** [comment]
I'll do my best to get to it later, meanwhile I'll simply point out that I
attempted to run xUnit tests using ReSharper with default parallelization
settings which means you get each test class running in parallel with other
test classes. All of them used BDDfy's fluent API, and when I ran 8 of them
all at once in a single ReSharper tests session some of them were failing
non-deterministicly due to weird dictionary concurrency issues. Is that
description enough or would you like me to upload fictive test classes to
the issue's discussion? Given the non-deterministic nature of threading
issues I think we should simply acknowledge there's a real problem with
mutable static state being used unprotected. That's also against .Net
standards - static methods must always be thread safe.

2016-01-31 16:57 GMT+02:00 Michael Whelan <notifications@github.com>:

**Comment by mwhelan** [comment]
That is helpful, but fictive test classes would be more concrete to work with! :-)

**Comment by Eldar1205** [comment]
Hi, thank you very much for fixing this :) 

When will a 4.2.1 version be released so this bug fix could reach BDDfy users?

**Comment by MehdiK** [comment]
This was released a while back. We just forgot to update this issue and close it off. Thanks for reporting @Eldar1205 and for your patience.
